### PR TITLE
Accuracy op

### PR DESCRIPTION
**Pull request by typhoonzero**
Fix https://github.com/PaddlePaddle/Paddle/issues/3840

### CODE REVIEW

File: paddle/operators/top_k_op.h [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    const size_t k = static_cast<int>(ctx.op().GetAttr<int>("k"));
+
+    T* output_data = output->mutable_data<T>(ctx.GetPlace());
+    int* indices_data = indices->mutable_data<int>(ctx.GetPlace());
+
+    auto X = EigenMatrix<T>::From(*input);
+
+    // reshape input to a flattern matrix(like flat_inner_dims)
+    framework::DDim inputdims = input->dims();
+    const size_t row = framework::product(
```
Comment (dzhwinter) [inline paddle/operators/top_k_op.h]: There will be a flatten_to_2d interface in ddim, maybe should put a TODO here to replace in the future.

File: paddle/operators/top_k_op.h [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        framework::slice_ddim(inputdims, 0, inputdims.size() - 1));
+    const size_t col = inputdims[inputdims.size() - 1];
+    Eigen::DSizes<int, 2> flat2dims(row, col);
+    X.reshape(flat2dims);
+
+    for (size_t i = 0; i < row; i++) {
+      // TODO(typhoonzero): make this more efficient
```
Comment (dzhwinter) [inline paddle/operators/top_k_op.h]: I don't think we have a better choice since the `partial_sort` is heap sort. 
`Topk` is such a classic question, so I don't think the efficiency is a problem.

File: paddle/operators/accuracy_op.cc [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+  AccuracyOpMaker(framework::OpProto *proto,
+                  framework::OpAttrChecker *op_checker)
+      : OpProtoAndCheckerMaker(proto, op_checker) {
+    AddInput("Inference", "topk(indices) the network output");
+    AddInput("Label", "Label of the training data");
+    // TODO(typhoonzero): AddInput("Weight", ...
+    AddOutput("Accuracy", "The accuracy of current batch");
+
+    AddComment(
+        R"DOC(Accuracy. It will print accuracy rate for classification.");
```
Comment (reyoung) [inline paddle/operators/accuracy_op.cc]: Bad comments here.

`R"DOC(...)DOC"` is just like python `"""..."""`, which `R"DOC(` is the left `"""` and `)DOC"` is the right `"""`.

Comment (lcy-seso) [reply paddle/operators/accuracy_op.cc]: Actually, I had been confused about what does `R"abc(....)abc"`  means.
I found this two doc solved my questions.

Just to share.
http://en.cppreference.com/w/cpp/language/string_literal
https://github.com/MicrosoftDocs/cpp-docs/blob/master/docs/cpp/string-and-character-literals-cpp.md

File: paddle/operators/top_k_op.h (hunk 1 above)
Comment (typhoonzero) [reply paddle/operators/top_k_op.h]: Also maybe we need different ways to flatten: `flatten_to_2d_inner` and `flatten_to_2d_outter`

File: paddle/operators/top_k_op.h (hunk 2 above)
Comment (typhoonzero) [reply paddle/operators/top_k_op.h]: I'm think try not to copy the memory, since the input is const, we cannot do it currently.

File: paddle/operators/accuracy_op.cu [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+ public:
+  void Compute(const framework::ExecutionContext& ctx) const override {
+    PADDLE_ENFORCE(platform::is_gpu_place(ctx.GetPlace()),
+                   "It must use GPUPlace.");
+    auto* inference = ctx.Input<Tensor>("Inference");
+    auto* label = ctx.Input<Tensor>("Label");
+    auto* accuracy = ctx.Output<Tensor>("Accuracy");
+    const int* inference_data = inference->data<int>();
+    const int* label_data = label->data<int>();
+    T* accuracy_data = accuracy->mutable_data<T>(ctx.GetPlace());
```
Comment (dzhwinter) [inline paddle/operators/accuracy_op.cu]: the accuracy_data should also be changed.

File: paddle/operators/accuracy_op.h [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        if (inference_data[i * class_dim + j] == label_data[i]) {
+          ++num_correct;
+          break;
+        }
+      }
+    }
+
+    // FIXME(typhoonzero): we don't accumulate the accuracy for now.
+    *accuracy_data =
+        static_cast<float>(num_correct) / static_cast<float>(num_samples);
```
Comment (dzhwinter) [inline paddle/operators/accuracy_op.h]: do we need an `ENFORCE` check num_samples is not equal to zero?  when user misuse this operator, will the num_samples be zero?  I'm not sure it's useful.

File: paddle/operators/accuracy_op.h [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+      for (size_t j = 0; j < class_dim; ++j) {
+        if (inference_data[i * class_dim + j] == label_data[i]) {
+          ++num_correct;
+          break;
+        }
+      }
+    }
+
+    // FIXME(typhoonzero): we don't accumulate the accuracy for now.
+    *accuracy_data = static_cast<T>(num_correct) / static_cast<T>(num_samples);
```
[the remark this answers is not in the document]
Comment (typhoonzero) [reply paddle/operators/accuracy_op.h]: Done.

File: paddle/operators/accuracy_op.h [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+template <typename Place, typename T>
+class AccuracyKernel : public framework::OpKernel {
+ public:
+  void Compute(const framework::ExecutionContext& ctx) const override {
+    auto* inference = ctx.Input<Tensor>("Inference");
+    auto* label = ctx.Input<Tensor>("Label");
+    auto* accuracy = ctx.Output<Tensor>("Accuracy");
+    const size_t topk = 1;
+    // static_cast<AttrType>(ctx.op_.GetAttr<AttrType>("topk"));
```
[the remark this answers is not in the document]
Comment (typhoonzero) [reply paddle/operators/accuracy_op.h]: Done.

File: paddle/operators/accuracy_op.h [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+using EigenScalar = framework::EigenScalar<T, MajorType, IndexType>;
+
+template <typename Place, typename T>
+class AccuracyKernel : public framework::OpKernel {
+ public:
+  void Compute(const framework::ExecutionContext& ctx) const override {
+    auto* inference = ctx.Input<Tensor>("Inference");
+    auto* label = ctx.Input<Tensor>("Label");
+    auto* accuracy = ctx.Output<Tensor>("Accuracy");
+    const size_t topk = 1;
```
[the remark this answers is not in the document]
Comment (typhoonzero) [reply paddle/operators/accuracy_op.h]: Removed.

File: paddle/operators/accuracy_op.cc [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+  void InferShape(const framework::InferShapeContext &ctx) const override {
+    PADDLE_ENFORCE_NOT_NULL(ctx.InputVar("Inference"),
+                            "Input of Inference must be initialized.");
+    PADDLE_ENFORCE_NOT_NULL(ctx.InputVar("Label"),
+                            "Input of Inference must be initialized.");
+    auto *inference = ctx.Input<framework::Tensor>("Inference");
+    auto *label = ctx.Input<framework::Tensor>("Label");
+
+    // label must be a vector
+    PADDLE_ENFORCE_EQ(label->dims().size(), 1);
```
[the remark this answers is not in the document]
Comment (typhoonzero) [reply paddle/operators/accuracy_op.cc]: Done.

File: paddle/operators/accuracy_op.cc [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+    PADDLE_ENFORCE_EQ(inference->dims()[0], label->dims()[0]);
```
[the remark this answers is not in the document]
Comment (typhoonzero) [reply paddle/operators/accuracy_op.cc]: Done.

File: paddle/operators/accuracy_op.cu (hunk 4 above)
Comment (typhoonzero) [reply paddle/operators/accuracy_op.cu]: Sorry, what do you mean by "change", accuracy_data is passed to cuda kernel and assign the result in the cuda kernel.

Comment (dzhwinter) [reply paddle/operators/accuracy_op.cu]: Just the same CPU code below.
```c++
float* accuracy_data = accuracy->mutable_data<float>(ctx.GetPlace());
```
Here the accuracy_data type is `T`, which may lose precision, maybe its better to write as `mutable_data<float>`. And I noticed that the AccuracyDivideKernel received a parameter of `float`.

File: paddle/operators/accuracy_op.cu [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+
+    size_t num_samples = inference->dims()[0];
+    size_t infer_width = inference->dims()[1];
+
+    AccuracyDivideKernel<<<1, 1>>>(num_samples, infer_width, 1, inference_data,
```
Comment (dzhwinter) [inline paddle/operators/accuracy_op.cu]: I'm not very familiar to cuda. It seems launched kernel in <<<1,1>>> can not utilize the capacity of cuda. This operator writes the similar one [crossEntropy](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/operators/cross_entropy_op.cu#L88), but I don't know how the `block=512` comes from.
@qingqing01

Comment (typhoonzero) [reply paddle/operators/accuracy_op.cu]: Well, I'm still trying to let the kernel use more threads, I'll enhance the kernel in next PR. I've got some problem writing kernel with `atocimAdd`.

File: paddle/operators/accuracy_op.h (hunk 5 above)
Comment (typhoonzero) [reply paddle/operators/accuracy_op.h]: Done, this return 0 if `num_sample==0`
